src/ore/src/ssh_key.rs

```diff
@@ -36,37 +39,139 @@ impl SshKeypair {
         });
 
-        let private_key_wrapper = PrivateKey::new(keypair, "materialize")?;
-        let openssh_private_key = &*private_key_wrapper.to_openssh(LineEnding::LF)?;
-        let openssh_public_key = private_key_wrapper.public_key().to_openssh()?;
+        let keypair = PrivateKey::new(keypair_data, "materialize")?;
+
+        Ok(SshKeypair { keypair })
+    }
+
+    /// Deserialize a keypair from a private key
+    ///
+    /// In the OpenSSH format, the public key is embedded inside the private key
+    fn from_private_key(private_key: &[u8]) -> Result<SshKeypair, anyhow::Error> {
+        let private_key = PrivateKey::from_openssh(private_key)?;
 
         Ok(SshKeypair {
-            public_key: openssh_public_key.as_bytes().to_vec(),
-            private_key: openssh_private_key.as_bytes().to_vec(),
+            keypair: private_key,
         })
     }
 
     /// Return the public key encoded in the OpenSSH format
-    pub fn ssh_public_key(&self) -> &[u8] {
-        &self.public_key
+    pub fn ssh_public_key(&self) -> String {
+        self.keypair.public_key().to_string()
     }
 
     /// Return the private key encoded in the OpenSSH format
-    pub fn ssh_private_key(&self) -> &[u8] {
-        &self.private_key
+    pub fn ssh_private_key(&self) -> Zeroizing<String> {
+        self.keypair.to_openssh(LineEnding::LF).expect("TODO")
+        // &self.private_key
+    }
+}
+
+impl PartialEq for SshKeypair {
+    fn eq(&self, other: &Self) -> bool {
+        self.keypair.fingerprint(Default::default())
+            == other.keypair.fingerprint(Default::default())
+    }
+}
+
+impl Eq for SshKeypair {}
+
+impl Serialize for SshKeypair {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("SshKeypair", 2)?;
+        // Public key is still encoded for backwards compatibility, but it is not used anymore
+        state.serialize_field("public_key", self.ssh_public_key().as_bytes())?;
+        state.serialize_field("private_key", self.ssh_private_key().as_bytes())?;
+        state.end()
+    }
+}
+
+use serde::de::{self, Deserializer, MapAccess, SeqAccess, Visitor};
+use zeroize::Zeroizing;
+
+impl<'de> Deserialize<'de> for SshKeypair {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(field_identifier, rename_all = "snake_case")]
+        enum Field {
+            PublicKey,
+            PrivateKey,
+        }
+
+        struct SshKeypairVisitor;
+
+        impl<'de> Visitor<'de> for SshKeypairVisitor {
+            type Value = SshKeypair;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("struct SshKeypair")
+            }
+
+            fn visit_seq<V>(self, mut seq: V) -> Result<SshKeypair, V::Error>
```

Comment:
did the derived impl actually generate a `visit_seq`?

Reply:
Good question, I don't know, but probably yes! From what I could read from the docs, `visit_seq` is used for some non-self describing formats (like Postcard) instead of `visit_map`. We are only using JSON here, which uses `visit_map`, but I preferred to err on the side of being more complete in case we want to change this in the future.